File: Distribution-Grid-EV-CA/R_Yanning/01_01_get_LDV_households_per_TAZ.R
library(data.table)

# ============== Short Distance Private Trips =================================
load.trips <- function() {
  files.all <- list.files()
  output <- data.table()
  for(f in files.all) {
    hold <- fread(f)
    output <- rbind(output,hold)
  }
  return(output)
}

setwd('C:/Nina/Research (large files)/CSTDM/SDPTM')
all.trips.sd <- load.trips()

TAZ.hh.sd.all <- all.trips.sd[Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3',
                               c('SerialNo','HomeZone')]
TAZ.hh.sd <- unique(TAZ.hh.sd.all)
setwd('C:/Users/Yanning Li/Box Sync/Distribution Grid/Distribution Grid EV CA')
fwrite(TAZ.hh.sd,file='data/mobility data/CSTDM processed/SDPTM_hh_home_TAZ.csv',row.names=FALSE)

# ============== Long Distance Private Trips =================================
all.trips.ld <- fread('C:/Nina/Research (large files)/CSTDM/LDPTM/LDPTM_Trips.csv')

TAZ.hh.ld.all <- all.trips.ld[Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3',
                              c('SerialNo','HomeZone')]
TAZ.hh.ld <- unique(TAZ.hh.ld.all)
fwrite(TAZ.hh.ld,file='data/mobility data/CSTDM processed/LDPTM_hh_home_TAZ.csv',row.names=FALSE)

all.trips.ld.accegr <- fread('C:/Nina/Research (large files)/CSTDM/LDPTM/LDPTM_AccEgr.csv')
nrow(all.trips.ld.accegr[is.na(AccMode)])
nrow(all.trips.ld.accegr[is.na(EgrMode)])

TAZ.hh.ld.accegr.all <- all.trips.ld.accegr[(DPurp=='Access'&AccMode=='Drive_Park')|(OPurp=='Egress'&EgrMode=='Rent_Car'),
                                            c('SerialNo','HomeZone')]
TAZ.hh.ld.accegr <- unique(TAZ.hh.ld.accegr.all)
fwrite(TAZ.hh.ld.accegr,file='data/mobility data/CSTDM processed/LDPTM_AccEgr_hh_home_TAZ.csv',row.names=FALSE)

# ============== ETM =================================
all.trips.etm <- fread('C:/Nina/Research (large files)/CSTDM/ETM/trips_Ext.csv')

# only the LDV trips that enter CA
TAZ.hh.etm.all <- all.trips.etm[(ActorType=='CarLong'|ActorType=='CarLocal')&(Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3') & DPurp=='I',
                                c('SerialNo','HomeZone')]

TAZ.hh.etm <- unique(TAZ.hh.etm.all)
fwrite(TAZ.hh.etm,file='data/mobility data/CSTDM processed/ETM_hh_home_TAZ.csv',row.names=FALSE)

# ============= start from here if all trips already processed ======================
setwd('C:/Users/Yanning Li/Box Sync/Distribution Grid/Distribution Grid EV CA')
TAZ.hh.sd <- fread('data/mobility data/CSTDM processed/SDPTM_hh_home_TAZ.csv')
TAZ.hh.ld <- fread('data/mobility data/CSTDM processed/LDPTM_hh_home_TAZ.csv')
TAZ.hh.ld.accegr <- fread('data/mobility data/CSTDM processed/LDPTM_AccEgr_hh_home_TAZ.csv')
TAZ.hh.etm <- fread('data/mobility data/CSTDM processed/ETM_hh_home_TAZ.csv')

# # Put the household pool inside CA together (ETM separately)
# TAZ.hh.CA.all <- rbind(TAZ.hh.sd,TAZ.hh.ld,TAZ.hh.ld.accegr)
# TAZ.hh.CA <- unique(TAZ.hh.CA.all)
# 
# fwrite(TAZ.hh.CA,file='data/mobility data/CSTDM processed/CA_hh_home_TAZ.csv',row.names=FALSE)

File: Distribution-Grid-EV-CA/01_01_get_LDV_households_per_TAZ.py
"""
01_01_get_LDV_households_per_TAZ.py — Python port of ``01_01_get_LDV_households_per_TAZ.R``.

Collects unique (SerialNo, HomeZone) pairs of LDV households from:
- CSTDM SDPTM (Short-Distance Passenger Travel Model)
- CSTDM LDPTM (Long-Distance Passenger Travel Model)  + AccEgr
- CSTDM ETM (External Travel Model)

Outputs per-source "hh home TAZ" CSVs.
"""

from __future__ import annotations

import os

import pandas as pd

from common import fread_all_in_dir

# ---------------------------------------------------------------------------
# Paths (mirror the R script)
# ---------------------------------------------------------------------------
WORK_ROOT = r"E:\GitHub\good_model\Distribution-Grid-EV-CA"

SDPTM_DIR = r"E:\Data\CSTDM_OD/SDPTM"
LDPTM_TRIPS = r"E:\Data\CSTDM_OD/LDPTM/LDPTM_Trips.csv"
LDPTM_ACCEGR = r"E:\Data\CSTDM_OD/LDPTM/LDPTM_AccEgr.csv"
ETM_TRIPS = r"E:\Data\CSTDM_OD/ETM/trips_Ext.csv"

OUT_DIR = r"E:\GitHub\good_model\Distribution-Grid-EV-CA\data\mobility_data\CSTDM_processed"
LDV_MODES = {"SOV", "HOV2", "HOV3"}


def main() -> None:
    os.makedirs(OUT_DIR, exist_ok=True)

    # --- Short Distance Private Trips ---------------------------------------
    sd = fread_all_in_dir(SDPTM_DIR)
    sd_hh = sd.loc[sd["Mode"].isin(LDV_MODES), ["SerialNo", "HomeZone"]]
    sd_hh = sd_hh.drop_duplicates()
    sd_hh.to_csv(os.path.join(OUT_DIR, "SDPTM_hh_home_TAZ.csv"), index=False)

    # --- Long Distance Private Trips ----------------------------------------
    ld = pd.read_csv(LDPTM_TRIPS)
    ld_hh = ld.loc[ld["Mode"].isin(LDV_MODES), ["SerialNo", "HomeZone"]].drop_duplicates()
    ld_hh.to_csv(os.path.join(OUT_DIR, "LDPTM_hh_home_TAZ.csv"), index=False)

    ld_ae = pd.read_csv(LDPTM_ACCEGR)
    print("NA AccMode:", ld_ae["AccMode"].isna().sum())
    print("NA EgrMode:", ld_ae["EgrMode"].isna().sum())

    mask_ae = ((ld_ae["DPurp"] == "Access") & (ld_ae["AccMode"] == "Drive_Park")) | (
        (ld_ae["OPurp"] == "Egress") & (ld_ae["EgrMode"] == "Rent_Car")
    )
    ld_ae_hh = ld_ae.loc[mask_ae, ["SerialNo", "HomeZone"]].drop_duplicates()
    ld_ae_hh.to_csv(os.path.join(OUT_DIR, "LDPTM_AccEgr_hh_home_TAZ.csv"), index=False)

    # --- ETM (only LDV trips that enter CA) ---------------------------------
    etm = pd.read_csv(ETM_TRIPS)
    mask_etm = (
        etm["ActorType"].isin(["CarLong", "CarLocal"])
        & etm["Mode"].isin(LDV_MODES)
        & (etm["DPurp"] == "I")
    )
    etm_hh = etm.loc[mask_etm, ["SerialNo", "HomeZone"]].drop_duplicates()
    etm_hh.to_csv(os.path.join(OUT_DIR, "ETM_hh_home_TAZ.csv"), index=False)


if __name__ == "__main__":
    main()
